File: services/inventory/internal/handlers/health.go
package handlers

import (
	"net/http"
)

// Healthz is a liveness probe. It never checks dependencies — it only
// confirms the process is up and able to handle HTTP requests.
func (h *Handler) Healthz(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// Readyz is a readiness probe. It actually checks PostgreSQL and Kafka
// connectivity, reporting 503 with a reason if either is unreachable.
func (h *Handler) Readyz(w http.ResponseWriter, r *http.Request) {
	if err := h.DB.Ping(r.Context()); err != nil {
		h.Logger.Warn("readiness check failed: postgres", errField(err))
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{
			"status": "not-ready",
			"reason": "postgres: " + err.Error(),
		})
		return
	}

	if h.KafkaCheck != nil {
		if err := h.KafkaCheck(); err != nil {
			h.Logger.Warn("readiness check failed: kafka", errField(err))
			writeJSON(w, http.StatusServiceUnavailable, map[string]string{
				"status": "not-ready",
				"reason": "kafka: " + err.Error(),
			})
			return
		}
	}

	writeJSON(w, http.StatusOK, map[string]string{"status": "ready"})
}

File: services/auth/src/routes/health.ts
import type { FastifyInstance } from 'fastify';
import { pingDatabase } from '../db/pool';
import { renderMetrics } from '../utils/metrics';

export async function healthRoutes(app: FastifyInstance): Promise<void> {
  // Liveness: no dependency checks, just "is the process up".
  app.get('/healthz', async (_request, reply) => {
    reply.code(200).send({ status: 'ok' });
  });

  // Readiness: must actually check downstream dependency connectivity (Postgres).
  app.get('/readyz', async (_request, reply) => {
    try {
      await pingDatabase();
      reply.code(200).send({ status: 'ready' });
    } catch (err) {
      app.log.warn({ err }, 'readiness check failed: database unreachable');
      reply.code(503).send({ status: 'not-ready', reason: 'database unreachable' });
    }
  });

  app.get('/metrics', async (_request, reply) => {
    reply.header('Content-Type', 'text/plain; version=0.0.4; charset=utf-8').send(renderMetrics());
  });
}
